{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true,
      "authorship_tag": "ABX9TyNY6GEHvTi0MR7MCbnDT152",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/AIKozyreva/masters-course-2024-genome-bioinformatics/blob/main/Lab_journal_hw1.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# –®–∞–≥ 1. –ü–æ–¥–≥–æ—Ç–æ–≤–∫–∞"
      ],
      "metadata": {
        "id": "i6Ir2K8jfcuk"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "mNAkzQ4oEW3O"
      },
      "outputs": [],
      "source": [
        "#–º–æ—Ä–∞–ª—å–Ω–∞—è"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#–≤—Å—Ç–∞–≤–∏—Ç—å –ø–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫–æ–Ω–¥—ã –∫ –∫–æ–ª–∞–±—É#"
      ],
      "metadata": {
        "id": "U1NxXiFGGDwO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install -q condacolab\n",
        "import condacolab\n",
        "condacolab.install()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bjlewwLC8MNa",
        "outputId": "39ee0f09-bc80-4005-9be1-d9183cdc1d70"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
            "\u001b[0m‚ú®üç∞‚ú® Everything looks OK!\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!conda version"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-ILzpN5Q9ww5",
        "outputId": "d9a6c701-f2c4-47b6-c8e1-ff830f276456"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "usage: conda [-h] [-v] [--no-plugins] [-V] COMMAND ...\n",
            "conda: error: argument COMMAND: invalid choice: 'version' (choose from 'activate', 'deactivate', 'clean', 'compare', 'config', 'create', 'info', 'init', 'install', 'list', 'notices', 'package', 'remove', 'uninstall', 'rename', 'run', 'search', 'update', 'upgrade', 'doctor', 'repoquery', 'env')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/035/770/615/GCF_035770615.1_bMelMel2.pri/GCF_035770615.1_bMelMel2.pri_genomic.fna.gz\n",
        "!wget https://hgdownload.soe.ucsc.edu/hubs/GCA/035/770/615/GCA_035770615.1/GCA_035770615.1.fa.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "mOrWNbM0Fi8R",
        "outputId": "628014e5-04ab-4a16-c8ba-462244724f81"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2024-03-04 21:05:56--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/035/770/615/GCF_035770615.1_bMelMel2.pri/GCF_035770615.1_bMelMel2.pri_genomic.fna.gz\n",
            "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.11, 130.14.250.12, 2607:f220:41e:250::13, ...\n",
            "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.11|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 423877926 (404M) [application/x-gzip]\n",
            "Saving to: ‚ÄòGCF_035770615.1_bMelMel2.pri_genomic.fna.gz.1‚Äô\n",
            "\n",
            "GCF_035770615.1_bMe 100%[===================>] 404.24M  32.8MB/s    in 10s     \n",
            "\n",
            "2024-03-04 21:06:06 (40.3 MB/s) - ‚ÄòGCF_035770615.1_bMelMel2.pri_genomic.fna.gz.1‚Äô saved [423877926/423877926]\n",
            "\n",
            "--2024-03-04 21:06:06--  http://./raw\n",
            "Resolving . (.)... failed: No address associated with hostname.\n",
            "wget: unable to resolve host address ‚Äò.‚Äô\n",
            "FINISHED --2024-03-04 21:06:06--\n",
            "Total wall clock time: 10s\n",
            "Downloaded: 1 files, 404M in 10s (40.3 MB/s)\n",
            "--2024-03-04 21:06:06--  https://hgdownload.soe.ucsc.edu/hubs/GCA/035/770/615/GCA_035770615.1/GCA_035770615.1.fa.gz\n",
            "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.198.53\n",
            "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.198.53|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 456876177 (436M) [application/x-gzip]\n",
            "Saving to: ‚ÄòGCA_035770615.1.fa.gz.1‚Äô\n",
            "\n",
            "GCA_035770615.1.fa. 100%[===================>] 435.71M  44.0MB/s    in 10s     \n",
            "\n",
            "2024-03-04 21:06:17 (42.2 MB/s) - ‚ÄòGCA_035770615.1.fa.gz.1‚Äô saved [456876177/456876177]\n",
            "\n",
            "--2024-03-04 21:06:17--  http://./raw\n",
            "Resolving . (.)... failed: No address associated with hostname.\n",
            "wget: unable to resolve host address ‚Äò.‚Äô\n",
            "FINISHED --2024-03-04 21:06:17--\n",
            "Total wall clock time: 11s\n",
            "Downloaded: 1 files, 436M in 10s (42.2 MB/s)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# –®–∞–≥ 2. –ò—Å–ø–æ–ª—å–∑—É–µ–º —Å–∫—Ä–∏–ø—Ç –¥–ª—è —Ç–æ–≥–æ, —á—Ç–æ–±—ã —Ä–∞—Å–ø–∞–∫–æ–≤–∞—Ç—å —Ñ–∞–π–ª—ã –∏ —Å—Ä–∞–∑—É –æ—Ü–µ–Ω–∏—Ç—å –∏—Ö —Å–æ–¥–µ—Ä–∂–∏–º–æ–µ - —Ä–∞–∑–º–µ—Ä—ã –∏ –∑–∞–≥–æ–ª–æ–≤–∫–∏."
      ],
      "metadata": {
        "id": "yys-A7ZafUob"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!ls"
      ],
      "metadata": {
        "id": "crVW7wsGGKwd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# —Å–∫—Ä–∏–ø—Ç - –Ω–∞–ø–∏—Å–∞–Ω –∑–∞—Ä–∞–Ω–µ–µ. out - —Ñ–∞–π–ª —Å–æ–∑–¥–∞—ë—Ç—Å—è, —á–µ—Ä–µ–∑ –∏—Ç–µ—Ä–∞—Ü–∏–∏ –î–û–∑–∞–ø–∏—Å—ã–≤–∞–µ—Ç—Å—è\n",
        "!./genInf_parser.sh\n",
        "!nano out_gen_info.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HoCs-LEoGvrc",
        "outputId": "dd34e60b-e351-4311-9929-38069b916212"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "gzip: /mnt/projects/users/aalayeva/genomics/raw/*.fa.gz: No such file or directory\n",
            "Decompressed: /mnt/projects/users/aalayeva/genomics/raw/*.fa.gz\n",
            "\\n\n",
            "gzip: /mnt/projects/users/aalayeva/genomics/raw/*.fna.gz: No such file or directory\n",
            "Decompressed: /mnt/projects/users/aalayeva/genomics/raw/*.fna.gz\n",
            "\\n\n",
            "gzip: /mnt/projects/users/aalayeva/genomics/raw/*.fastq.gz: No such file or directory\n",
            "Decompressed: /mnt/projects/users/aalayeva/genomics/raw/*.fastq.gz\n",
            "\\n\n",
            "gzip: /mnt/projects/users/aalayeva/genomics/raw/*.fasta.gz: No such file or directory\n",
            "Decompressed: /mnt/projects/users/aalayeva/genomics/raw/*.fasta.gz\n",
            "\\n\n",
            "/bin/bash: line 1: nano: command not found\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#–≤—Å–ø–æ–º–æ–≥–∞—Ç–µ–ª—å–Ω–∞—è –Ω–∞ —Å–ª—É—á–∞–π –ø—Ä–æ–±–ª–µ–º –ø—Ä–∏ –ø–µ—Ä–µ–Ω–æ—Å–µ —Ñ–∞–π–ª–æ–≤ —Å –≤–∏–Ω–¥—ã –Ω–∞ linux (–µ—Å–ª–∏ –ø—Ä–µ–¥.–∫–æ–º–∞–Ω–¥–∞ –Ω–µ —Å—Ä–∞–±–∞—Ç—ã–≤–∞–µ—Ç, —Ç–æ —Å–Ω–∞—á–∞–ª–∞ —ç—Ç—É —Å–¥–µ–ª–∞—Ç—å)\n",
        "!conda install dos2unix\n",
        "! dos2unix genInf_parser.sh\n",
        "!chmod +x genInf_parser.sh"
      ],
      "metadata": {
        "id": "9nlBvN6UGyPy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "–ü–æ–ª—É—á–∞–µ–º –≤ –≤—ã–æ–¥–µ —á—Ç–æ-—Ç–æ —Ç–∞–∫–æ–µ\n",
        "\n",
        "Processing GCA_035770615.1.fa\n",
        "path to file is /mnt/projects/users/aalayeva/genomics/raw/GCA_035770615.1.fa\n",
        "File size is 1.5G\n",
        "GCA_035770615.1.fa amount of lines: 30825768\n",
        "GCA_035770615.1.fa amount of headers: 354\n",
        "GCA_035770615.1.fa header information: CM069555.1\n",
        "\n",
        "Processing GCF_035770615.1_bMelMel2.pri_genomic.fna\n",
        "path to file is /mnt/projects/users/aalayeva/genomics/raw/GCF_035770615.1_bMelMel2.pri_genomic.fna\n",
        "File size is 1.5G\n",
        "GCF_035770615.1_bMelMel2.pri_genomic.fna amount of lines: 19266095\n",
        "GCF_035770615.1_bMelMel2.pri_genomic.fna amount of headers: 353\n",
        "GCF_035770615.1_bMelMel2.pri_genomic.fna header information: NC_086194.1 Melospiza melodia melodia isolate bMelMel2 chromosome 1, bMelMel2.pri, whole genome shotgun sequence"
      ],
      "metadata": {
        "id": "BGOO-Hz9Hscu"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# –®–∞–≥ 3. –ü—Ä–æ–≤–µ—Ä–∫–∞ –∫–∞—á–µ—Å—Ç–≤–∞ by QUAST"
      ],
      "metadata": {
        "id": "JicUdNSvHhws"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "> –ü–∞—Ä–∞–º–µ—Ç—Ä—ã\n",
        "\n",
        "```\n",
        "--eukaryote (or -e)\n",
        "Genome is eukaryotic. Affects everything:\n",
        "For default, GeneMarkS is used. For eukaryotes, GeneMark-ES is used.\n",
        "Barrnap will use eukaryotic database to predict ribosomal RNA genes.\n",
        "\n",
        "By default, QUAST assumes that a genome is circular and correctly processes its linear representation. This options indicates that the genome is not circular.\n",
        "\n",
        "--large\n",
        "Genome is large (typically > 100 Mbp). Use optimal parameters for evaluation of large genomes. Affects speed and accuracy. In particular, imposes --eukaryote --min-contig 3000 --min-alignment 500 --extensive-mis-size 7000.\n",
        "\n",
        "--fast\n",
        "A shortcut for using all of speedup options except --no-check.\n",
        "\n",
        "--silent\n",
        "Do not print detailed information about each step in standard output.\n",
        "```\n",
        "\n"
      ],
      "metadata": {
        "id": "f_1TZnfpLEye"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!conda install -c bioconda quast"
      ],
      "metadata": {
        "id": "hJBVz2SH94_i"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!quast"
      ],
      "metadata": {
        "id": "KyjNYVQ6981d"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!quast.py --eukaryote --large --fast --silent /content/GCA_035770615.1.fa"
      ],
      "metadata": {
        "id": "L0lCkBrcLBtX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# –®–∞–≥ 4. –ó–∞–ø—É—Å–∫ BUSCO"
      ],
      "metadata": {
        "id": "1I_d5XKMHQr7"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "`–ö–æ–º–∞–Ω–¥–∞: *busco -i /*.fa -m genome -l /*_odb10 --offline -f --download_path ./busco_downloads *`\n",
        "\n",
        "\n",
        "> –ü–∞—Ä–∞–º–µ—Ç—Ä—ã\n",
        "\n",
        "> -l -> —É–∫–∞–∑–∞–Ω–∏–µ –ª–æ–∫–∞–ª—å–Ω–æ–π –±–∞–∑—ã –¥–∞–Ω–Ω—ã—Ö\n",
        "\n",
        "> -f ->–ø–µ—Ä–µ–ø–∏—Å—ã–≤–∞–Ω–∏–µ —Å—É—â–µ—Å—Ç–≤—É—é—â–µ–π –ø–∞–ø–∫–∏ –∑–∞–ø—É—Å–∫–∞, —Å—É—â–µ—Å—Ç–≤—É—é—â–µ–π —Å –Ω–µ—É–¥–∞—á–Ω–æ–≥–æ —Ä–∞–∑–∞\n",
        "\n",
        "> -offline -> —á—Ç–æ–±—ã –æ–Ω –Ω–∏—á–µ–≥–æ –¥–æ–ø–æ–ª–Ω–∏—Ç–µ–ª—å–Ω–æ –Ω–µ –ø—ã—Ç–∞–ª—Å—è —Å–∫–∞—á–∏–≤–∞—Ç—å, –ø–æ—Ç–æ–º—É —á—Ç–æ –∏–Ω–∞—á–µ –æ–Ω –≤–∏—Å–∞–Ω—ë—Ç –∏–∑-–∑–∞ –Ω–µ–≤–æ–∑–º–æ–∂–Ω–æ—Å—Ç–∏ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –≤–ø–Ω ü§°üëç\n",
        "\n",
        "> --download_path -> —è –Ω–∞ –≤—Å—è–∫–∏–π —Å–ª—É—á–∞–π –ø—Ä–æ–ø–∏—Å–∞–ª–∞, –ø–æ—Ç–æ–º—É —á—Ç–æ –≤—Ä–æ–¥–µ –∫–∞–∫ –æ–±—ã—á–Ω–æ –±–∞–∑—ã –ª–µ–∂–∞—Ç –≤ —ç—Ç–æ–π –ø–∞–ø–∫–µ, –∞ —É –º–µ–Ω—è –æ–Ω–∏ –Ω–µ –≤ –Ω–µ–π, –∞ –Ω–∞ –æ–¥–Ω—É –ø–∞–ø–∫—É –≤—ã—à–µ, –ø—Ä–æ—Å—Ç–æ –≤ busco/\n",
        "\n",
        "> -m -> —ç—Ç–æ —É–∫–∞–∑–∞–Ω–∏–µ —Ä–µ–∂–∏–º–∞ —Ä–∞–±–æ—Ç—ã, —è –≥–µ–Ω–æ–º —Å–æ–±–∏—Ä–∞—é, –ø–æ—ç—Ç–æ–º—É –≥–µ–Ω–æ–º\n",
        "\n",
        "> -i -> —ç—Ç–æ –≤—Ö–æ–¥–Ω–∞—è —Å—ã—Ä–∞—è .fasta –Ω–∞ –≤—Ö–æ–¥\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "_sI3YPlSA9mG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#c–æ–∑–¥–∞—ë–º –æ–∫—Ä—É–∂–µ–Ω–∏–µ –∏ —Å–∫–∞—á–∏–≤–∞–µ–º –±—É—Å–∫–æ, –¥–∞ - –∏–Ω–∞—á–µ –æ–Ω –Ω–µ —É—Å—Ç–∞–Ω–æ–≤–∏—Ç—Å—è\n",
        "!conda create -n busco -c conda-forge -c bioconda busco=5.6.1"
      ],
      "metadata": {
        "id": "eaAc1jTOEwNV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#–≤–∫–ª—é—á–∞–µ–º vpn –≤ –±—Ä–∞—É–∑–µ—Ä–µ –∏ –∏–¥—ë–º –Ω–∞ —Å–∞–π—Ç –Ω–∏–∂–µ\n",
        "# https://busco-data.ezlab.org/v5/data/lineages/\n",
        "#–∫–∞—á–∞–µ–º –Ω—É–∂–Ω—É—é –Ω–∞–º –ë–î –≤ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏ –æ—Ç –æ–±—ä–µ–∫—Ç–∞ —Ä–∞–±–æ—Ç—ã"
      ],
      "metadata": {
        "id": "5d9dJFWXW0yb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#–≤–æ–∑–≤—Ä–∞—â–∞–µ–º—Å—è –Ω–∞ —Å–µ—Ä–≤–µ—Ä, –∑–∞–≥—Ä—É–∂–∞–µ–º –Ω–∞ —Å–µ—Ä–≤–µ—Ä —Å–≤–æ—é –ë–î"
      ],
      "metadata": {
        "id": "xYJ026tBXbMv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#—Ä–∞—Å–ø–∞–∫–æ–≤—ã–≤–∞–µ–º —Ñ–∞–π–ª –±–¥, –æ–Ω –ø—Ä–µ–≤—Ä–∞—Ç–∏—Ç—Å—è –≤ –ø–∞–ø–∫—É —Å –±–¥"
      ],
      "metadata": {
        "id": "Oegbtlo1X9mq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!tar -xzf /mnt/projects/users/aalayeva/genomics/busco/busco_downloads/passeriformes_odb10.2024-01-08.tar.gz"
      ],
      "metadata": {
        "id": "ZXnmBSm1XZQi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#—Ç–µ–ø–µ—Ä—å –º–æ–∂–Ω–æ –ø—Ä–æ–ø–∏—Å—ã–≤–∞—Ç—å –∫–æ–º–∞–Ω–¥—É –±—É—Å–∫–æ"
      ],
      "metadata": {
        "id": "PXuqk124X8kX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!busco -i /mnt/projects/users/aalayeva/genomics/raw/GCA_035770615.1.fa -m genome -l /mnt/projects/users/aalayeva/genomics/busco/passeriformes_odb10 --offline --download_path /mnt/projects/users/aalayeva/genomics/busco/busco_downloads"
      ],
      "metadata": {
        "id": "3HP-RRqlX6xU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "wvXeADYMYKYo"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}